docs/jsonapi-spec.md

```diff
@@ -591,6 +592,83 @@ See [Multi-Document Failure Considerations](#multi-document-failure-consideratio
 
 If an error occurs the command will not return `data`.
 
+### findOneAndReplace Command
+
+`findOneAndReplace` replaces a single document based on the specified filter clause. 
+
+*Syntax:*
+
+```bnf
+<find-one-and-replace-command> ::= findOneAndReplace <find-one-and-replace-command-payload> 
+<find-one-and-replace-command-payload> ::= <filter-clause>? 
+                             <sort-clause>? 
+                             <replacement-clause>?
+                             <find-one-and-replace-command-options>?
+
+<find-one-and-replace-command-options> ::= (<find-option-name>)*
+```
+
+*Sample:*
+
+```json5
+// Replaces a single document based on the specified filter and sort. Returns a document in response.
+{ "findOneAndReplace" : 
+    { "filter" : {<filter-clause>}, 
+      "sort" : {<sort-clause>}, 
+      "replacement" : {<document-to-replace>}, 
+      "options" : 
+          {"upsert" : true/false, 
```

Comment:
I don't think `"upsert"` option is accepted.

Reply:
However, not 100% sure if "upsert" SHOULD be supported and is just missing currently?

/cc @maheshrajamani

Reply:
This is my mistake. I put it wrongly in the parent issue. The only option support for this command is `returnDocument`.

Reply:
Will fix the spec

Reply:
Any particular reason why we wouldn't support upsert for findOneAndReplace? Do we support upsert for findOneAndUpdate?

Reply:
We do support `upsert` for `findOneAndUpdate`.  I did notice that the DB-we-dont-name also allows it for `findOneAndReplace` options, not sure why we wouldn't want it. But right now our implementation does not (nor spec).

Reply:
@vkarpov15 The upsert logic currently adds all equality filter fields and update operation. In case of replace this logic doesn't make sense to add filter fields and later replace it with fields in replacement document. Any thoughts on this so i can have it implemented?

Reply:
We do not necessarily have to implement if it does not make sense @maheshrajamani. Sounds like we should not try to implement it at this point, as per good points you raise.

Reply:
@maheshrajamani `findOneAndReplace()` with upsert just uses the replacement document, doesn't apply the query fields. See the below example from MongoDB shell:

```
MongoDB Enterprise > db.tests.findOneAndReplace({ name: 'foo', email: 'foo@bar.baz' }, { name: 'bar', age: 42 }, { upsert: true })
null
MongoDB Enterprise > db.tests.findOne()
{ "_id" : ObjectId("642dbbc72dfa2abc7d1319e2"), "name" : "bar", "age" : 42 }
MongoDB Enterprise > 
```

Like @tatu-at-datastax , `upsert` support for `findOneAndReplace()` is not terribly important currently.

Reply:
Updated - thanks